### PR TITLE
feat(zb_io): support external taps + casks with safer install/uninstall behavior

### PR DESCRIPTION
**Pull request by maria-rcks**
- [x] I have run all relevant linters for this PR.
- [x] I have read and will follow the contributing guidelines.

#

- [x] AI was used in this PR to generate code, documentation, or other changes, and I have explained _how_ it was used below. I understand that if I do not adhere to disclosing AI usage, my PR can be closed, even without explanation.
- [ ] I have not used AI for _any_ portion of this PR.

  changes:
  - adds external tap formula support (`owner/tap/formula`) with ruby parser-based bottle resolution
  - adds cask install support with zip extraction + binary artifact staging
  - adds cleanup/rollback path for cask install failures (link/db failure cases)
  - hardens cask source path handling (rejects absolute path + `..` traversal)
  - fixes uninstall identity edge case so uninstalling a non-installed tap ref does not remove an installed core formula with same name
  - improves tap bottle parsing for nested `do/end` blocks
  - adds regression tests for uninstall identity collisions and cask path/rollback behavior

  validation:
  - `cargo fmt`
  - `just lint`
  - `just test`
  - e2e checked with fresh root/prefix for:
    - cask install/list/uninstall/list
    - tap install/list/uninstall/list
    - negative case: `uninstall hashicorp/tap/terraform` does not remove installed core `terraform`

  > AI-assisted: planning + initial implementation with gpt-5.3-codex, manually reviewed and tested.

### CODE REVIEW

**Comment by maria-rcks** [comment]
\*needs macos testing

**Comment by mac9sb** [comment]
Any update on this one?

**Comment by cachebag** [comment]
> Any update on this one?

Reviewing soon. @maria-rcks resolve conflicts?

**Comment by maria-rcks** [comment]
sorry, will do! i've been busy with the website

File: zb_io/src/network/api.rs [hunk 1]
```diff
@@ -105,6 +126,112 @@ impl ApiClient {
...
+            ),
+        ];
+        let branches = vec!["HEAD", "main", "master"];
+
+        let mut last_status: Option<reqwest::StatusCode> = None;
+
+        for repo in candidate_repos {
+            for branch in &branches {
+                for path in &candidate_paths {
+                    let url = format!(
```
Comment (cachebag) [inline zb_io/src/network/api.rs]: This is insane. We're firing up to 12 sequential HTTP requests to find a single formula. That's just brutal.

Remove "HEAD" from branches (it's redundant with main on raw.githubusercontent.com and may not even resolve), and consider `try_join!`-ing at least the path candidates in parallel.

File: zb_io/src/installer/install.rs [hunk 2]
```diff
@@ -236,11 +240,14 @@ impl Installer {
...
-        let to_install: Vec<(Formula, SelectedBottle)> = plan
-            .formulas
+        // Pair install identities with formulas and bottles.
+        // `install_name` is persisted to DB; `formula.name` is used for bottle layout/linking.
+        let to_install: Vec<(String, Formula, SelectedBottle)> = plan
+            .install_names
             .into_iter()
+            .zip(plan.formulas.into_iter())
             .zip(plan.bottles.into_iter())
+            .map(|((install_name, formula), bottle)| (install_name, formula, bottle))
```
Comment (cachebag) [inline zb_io/src/installer/install.rs]: Three parallel vecs zipped together is fragile. `.zip()` silently truncates if lengths differ. I thnk that `debug_assert_eq!(plan.install_names.len(), plan.formulas.len())` would be cheap insurance here..

Or honestly even better, make `InstallPlan` hold a `Vec<(String, Formula, SelectedBottle)>` directly.

File: zb_io/src/installer/cask.rs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        install_name: format!("cask:{token}"),
+        token: token.to_string(),
+        version,
+        url,
+        sha256,
+        binaries,
+    })
+}
+
+fn required_string(value: &Value, field: &str) -> Result<String, Error> {
```
Comment (cachebag) [inline zb_io/src/installer/cask.rs:57]: Returns `Error::NetworkFailure` for really is JSON schema/parse errors. We should swap to `InvalidArgument`.

File: zb_io/src/installer/cask.rs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+            "tahoe", "sequoia", "sonoma", "ventura", "monterey", "big_sur", "catalina",
+        ]
+    }
+    #[cfg(not(any(target_os = "linux", target_os = "macos")))]
+    {
+        &[]
+    }
+}
+
+fn parse_binary_artifacts(cask: &Value) -> Result<Vec<CaskBinary>, Error> {
```
Comment (cachebag) [inline zb_io/src/installer/cask.rs:106]: Same here

https://github.com/lucasgelfond/zerobrew/pull/203/changes#r2791291266

File: zb_io/src/installer/install.rs [hunk 5]
```diff
@@ -520,6 +580,204 @@ impl Installer {
     pub fn keg_path(&self, name: &str, version: &str) -> std::path::PathBuf {
         self.cellar.keg_path(name, version)
     }
+    async fn install_single_cask(&mut self, token: &str, link: bool) -> Result<(), Error> {
```
Comment (cachebag) [inline zb_io/src/installer/install.rs:583]: This call-sites for this are super undesirable. Scope-guard (e.g., scopeguard::defer! or a small RAII wrapper) would collapse all of that into one declaration at the top, making the happy path much more readable.

File: zb_io/src/network/tap_formula.rs
Comment (cachebag) [inline zb_io/src/network/tap_formula.rs:1]: Hmm I don't really like that every call to `parse_version`, `parse_revision`, `parse_dependencies`, `parse_bottle_files`, and `extract_bottle_block` recompiles its regexes. This will smother us with dep trees. Use something like `LazyLock<Regex>`. Easy win

File: zb_io/src/network/tap_formula.rs [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+        let Some(tag) = cap.get(1).map(|m| m.as_str()) else {
+            continue;
+        };
+        let Some(sha) = cap.get(2).map(|m| m.as_str()) else {
+            continue;
+        };
+        if matches!(tag, "cellar") {
+            continue;
+        }
+        let url = build_bottle_url(spec, root_url, stable, revision, rebuild, tag, sha);
```
Comment (cachebag) [inline zb_io/src/network/tap_formula.rs:233]: Best you just do `tag == "cellar"` here

File: zb_io/src/installer/install.rs [hunk 7]
```diff
@@ -520,6 +580,204 @@ impl Installer {
...
+                link,
+            );
+            return Err(e);
+        }
+
+        Ok(())
+    }
+
+    fn stage_cask_binaries(
+        &self,
```
Comment (cachebag) [inline zb_io/src/installer/install.rs]: I don't see `&self` being used? Should either be a free function like `resolve_cask_source_path` already is. Making it freestanding would also make it easier to unit test in isolation.

File: zb_io/src/network/tap_formula.rs
Comment (cachebag) [inline zb_io/src/network/tap_formula.rs:1]: Error handling here is also messy. I know we don't have anything solidified yet but in this context specifically, it makes no sense to switch between `ok()?`,  and `expect()`

Unify these with `expect()`. It's more honest- `ok()?` jsut  silently propagates `None` for an impossible failure which will mask bugs eventually upstream

File: zb_io/src/network/tap_formula.rs [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+    for cap in dep_re.captures_iter(source) {
+        let options = cap.get(2).map(|m| m.as_str()).unwrap_or("");
+        if options.contains(":build") || options.contains(":test") {
+            continue;
+        }
+        if let Some(dep) = cap.get(1) {
+            deps.push(dep.as_str().to_string());
+        }
+    }
+    deps.sort();
```
Comment (cachebag) [inline zb_io/src/network/tap_formula.rs]:
```suggestion
    deps.sort_unstable();
```

There's no reason to preserve relative order of equal elements on `Vec<String>`

File: zb_io/src/installer/install.rs [hunk 9]
```diff
@@ -459,8 +490,37 @@ impl Installer {
...
+    }
+
+    pub async fn install_casks(
+        &mut self,
+        names: &[String],
+        link: bool,
+    ) -> Result<ExecuteResult, Error> {
+        let mut installed = 0usize;
+        for name in names {
+            let token = name.strip_prefix("cask:").unwrap_or(name);
```
Comment (cachebag) [inline zb_io/src/installer/install.rs]: `expect()` here instead, please.

**Review by cachebag — CHANGES_REQUESTED** [review]

**Comment by maria-rcks** [comment]
@cachebag if there's anything missing lmk (sorry for the bad practices at the start)

File: zb_io/src/installer/install.rs (hunk 2 above)
Comment (cachebag) [reply zb_io/src/installer/install.rs]: There is still no `debug_assert_eq!` added to guard against length mismatches, and `InstallPlan` is not changed to store a `Vec<(String, Formula, SelectedBottle)>` directly.

Unless I'm missing something?

File: zb_io/src/network/tap_formula.rs (hunk 6 above)
Comment (cachebag) [reply zb_io/src/network/tap_formula.rs:233]: Still not addressed?

**Comment by maria-rcks** [comment]
I think github is not showing you the latest commits on the changes tab, try and see if its set to:

image
